23: Breaks up nearest tests to not be singular test

khurd21 opened the pull request:
Tests modified:
- floor_test.cpp
- trunc_test.cpp

Both adopts parameterized tests. These help reduce the expect calls and code duplication, also making it easier to add additional test cases in the future. NaN tests are still inside of the original TEST function.

Since the tests seemed to be directly comparing the function under test with the standard library equivalent, adjusted the `std::signbit` and `std::isnan` comparisons to directly compare equality in the NaN tests.

I also made sure to run the `clang-format` on the files that were modified.

<!-- Please target the "dev" branch when creating a pull request to ensure you are using the latest version of ccmath -->

khurd21 commented:
I saw this project and thought it was very well documented and heavily tested! I wanted to use this PR as an introduction and to see if changes like this for the tests are okay! I saw your card #43. If these changes look okay, I would love to look into the other ones and format them similarly.

Rinzii commented:
Hello! Firstly, thank you for taking the time to create a PR for ccmath! 

Overall, these changes are perfectly acceptable the only request I would have is that somewhere in the test we perform a `static_assert` to ensure the functions are being evaluated at compile time, but overall these changes seem pretty good!

Rinzii reviewed (changes requested):
All required at the moment would be to add a static_assert to both of these test cases to ensure our functions work at compile time. Besides lgtm!

khurd21 commented:
I tried giving this a good tackle, but I noticed the ninja-clang builds were failing. My first question, I have a linux vm on my computer, is there a way I can set up the ninja clang to try and replicate this locally? It looks like there might be an issue with compile time for truncl, as adding a static assert for that test yielded this error.

I tried compiling with clang on linux on the master branch and also saw these errors upon trying to build:

```sh
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/support/type_traits.hpp:101:21: error: redefinition of 'internal_is_floating_point_helper<long double>'
        template <> struct internal_is_floating_point_helper<types::float128> : std::true_type{};
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/support/type_traits.hpp:84:21: note: previous definition is here
        template <> struct internal_is_floating_point_helper<long double> : std::true_type{};
                           ^
In file included from /home/ubuntu/Workplace/ccmath/example/main.cpp:9:
In file included from /home/ubuntu/Workplace/ccmath/include/ccmath/ccmath.hpp:12:
In file included from /home/ubuntu/Workplace/ccmath/include/ccmath/basic.hpp:13:
In file included from /home/ubuntu/Workplace/ccmath/include/ccmath/math/basic/fma.hpp:14:
In file included from /home/ubuntu/Workplace/ccmath/include/ccmath/math/compare/signbit.hpp:21:
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/support/bits.hpp:262:10: warning: implicit conversion changes signedness: 'unsigned int' to 'std::enable_if_t<traits::ccm_is_unsigned_v<unsigned __int128>, int>' (aka 'int') [-Wsign-conversion]
                return zero_bits;
                ~~~~~~ ^~~~~~~~~
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/generic/functions/power/sqrt_gen.hpp:47:31: note: in instantiation of function template specialization 'ccm::support::countl_zero<unsigned __int128>' requested here
                        const int shift = support::countl_zero(mantissa) - (8 * static_cast<int>(sizeof(mantissa)) - 1 - support::fp::FPBits<T>::fraction_length);
                                                   ^
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/generic/functions/power/sqrt_gen.hpp:184:16: note: in instantiation of function template specialization 'ccm::gen::internal::normalize<long double>' requested here
                                        internal::normalize<T>(x_exp, x_mant);
                                                  ^
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/generic/functions/power/sqrt_gen.hpp:279:13: note: in instantiation of function template specialization 'ccm::gen::internal::impl::sqrt_calc_bits<long double>' requested here
                                        return sqrt_calc_bits(bits);
                                               ^
/home/ubuntu/Workplace/ccmath/include/ccmath/internal/generic/functions/power/sqrt_gen.hpp:288:26: note: in instantiation of function template specialization 'ccm::gen::internal::impl::sqrt_impl<long double>' requested here
                return internal::impl::sqrt_impl(num);
                                       ^
/home/ubuntu/Workplace/ccmath/include/ccmath/math/power/sqrt.hpp:38:65: note: in instantiation of function template specialization 'ccm::gen::sqrt_gen<long double, true>' requested here
                if (ccm::support::is_constant_evaluated()) { return ccm::gen::sqrt_gen<T>(num); }
                                                                              ^
/home/ubuntu/Workplace/ccmath/include/ccmath/math/power/sqrt.hpp:72:15: note: in instantiation of function template specialization 'ccm::sqrt<long double, true>' requested here
                return ccm::sqrt<long double>(num);
                            ^
1 warning and 1 error generated.
gmake[2]: *** [example/CMakeFiles/ccmath-examples.dir/build.make:76: example/CMakeFiles/ccmath-examples.dir/main.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:339: example/CMakeFiles/ccmath-examples.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

Do you have any familiarity with this?

Thanks!

Rinzii commented:
> I tried giving this a good tackle and think I stumbled upon an issue. I noticed when I added the tests for compile time assertions, the`ccm::truncl` yieled an error for the ninja-clang build only. The error is here:
> 
> https://github.com/Rinzii/ccmath/actions/runs/9848897543/job/27191697093#step:5:82
> 
> It seems like there is an improper type being passed into `__builtin_bit_cast` based on the message. I was unable to replicate this locally. I attempted to use the ninja-clang configuration preset but I was unable to get a successful build yet.
> 
> In the most recent push, I commented out the internals of the test with the `constexpr ccm::truncl` and it seems there is no issue with any of the other functions in the nearest folder.
> 
> What are you thoughts on this?

Sorry I‘ve been having a lot of IRL stuff going on recently and have not had much time to look at my GitHub. Is this issue still persisting?

khurd21 commented:
> > I tried giving this a good tackle and think I stumbled upon an issue. I noticed when I added the tests for compile time assertions, the`ccm::truncl` yieled an error for the ninja-clang build only. The error is here:
> > https://github.com/Rinzii/ccmath/actions/runs/9848897543/job/27191697093#step:5:82
> > It seems like there is an improper type being passed into `__builtin_bit_cast` based on the message. I was unable to replicate this locally. I attempted to use the ninja-clang configuration preset but I was unable to get a successful build yet.
> > In the most recent push, I commented out the internals of the test with the `constexpr ccm::truncl` and it seems there is no issue with any of the other functions in the nearest folder.
> > What are you thoughts on this?
> 
> Sorry I‘ve been having a lot of IRL stuff going on recently and have not had much time to look at my GitHub. Is this issue still persisting?

No problem. Yes, the issue still is persisting. I have been trying to build using clang on Ubuntu locally, however I was getting other build errors before this one. In the current commit, I have commented out the compile time test I added for `ccm::truncl`. It seems that `truncl` is the only function out of the `nearest` tests that does not work at compile time for Linux Ninja Clang build.

If this sounds like an okay plan, I leave commented out compile time test [here](https://github.com/Rinzii/ccmath/pull/43/files#diff-0ad15176bfaf60d1dd519b442c7412c099c1d2425808bd30e4823cf9ddd4722bR191) and have a card to address this properly in the future?

Rinzii commented:
> > > I tried giving this a good tackle and think I stumbled upon an issue. I noticed when I added the tests for compile time assertions, the`ccm::truncl` yieled an error for the ninja-clang build only. The error is here:
> > > https://github.com/Rinzii/ccmath/actions/runs/9848897543/job/27191697093#step:5:82
> > > It seems like there is an improper type being passed into `__builtin_bit_cast` based on the message. I was unable to replicate this locally. I attempted to use the ninja-clang configuration preset but I was unable to get a successful build yet.
> > > In the most recent push, I commented out the internals of the test with the `constexpr ccm::truncl` and it seems there is no issue with any of the other functions in the nearest folder.
> > > What are you thoughts on this?
> > 
> > 
> > Sorry I‘ve been having a lot of IRL stuff going on recently and have not had much time to look at my GitHub. Is this issue still persisting?
> 
> No problem. Yes, the issue still is persisting. I have been trying to build using clang on Ubuntu locally, however I was getting other build errors before this one. In the current commit, I have commented out the compile time test I added for `ccm::truncl`. It seems that `truncl` is the only function out of the `nearest` tests that does not work at compile time for Linux Ninja Clang build.
> 
> If this sounds like an okay plan, I leave commented out compile time test [here](https://github.com/Rinzii/ccmath/pull/43/files#diff-0ad15176bfaf60d1dd519b442c7412c099c1d2425808bd30e4823cf9ddd4722bR191) and have a card to address this properly in the future?

Hmmm I’ll investigate this issue myself locally in a few days. I’ll see if I can identify what the problem may be. :)

Rinzii commented:
Ok, after looking into this it appears the issue is because of specific implementation details of how `__builtin_bit_cast` works on clang vs GCC. This issue is appearing due to `__builtin_bit_cast` inability to determine the value of long double since it will always be 16 bytes, but can be represented as either 80 or 128 bits. GCC appears to just let it fly but clang does not and errors out. There is no real simple way to solve this issue, so for simplicity I'd just get rid of this test with a macro if we are on clang and testing `truncl` specifically.

khurd21 commented:
> Ok, after looking into this it appears the issue is because of specific implementation details of how `__builtin_bit_cast` works on clang vs GCC. This issue is appearing due to `__builtin_bit_cast` inability to determine the value of long double since it will always be 16 bytes, but can be represented as either 80 or 128 bits. GCC appears to just let it fly but clang does not and errors out. There is no real simple way to solve this issue, so for simplicity I'd just get rid of this test with a macro if we are on clang and testing `truncl` specifically.

Sounds good! I updated the code. I added a macro to the cpp file here: https://github.com/Rinzii/ccmath/pull/43/files#diff-0ad15176bfaf60d1dd519b442c7412c099c1d2425808bd30e4823cf9ddd4722bR17

Since clang linux was the only one failing, I made sure to check the compiler was clang and the os was linux. I then undefined it at the end of the source file to try and reduce the scope of the macro to be just that test file.

Rinzii reviewed (approved):
lgtm!
